include/aws/common/message_bus.h

```diff
@@ -0,0 +1,78 @@
+#ifndef AWS_COMMON_MESSAGE_BUS_H
+#define AWS_COMMON_MESSAGE_BUS_H
+
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/common.h>
+#include <aws/common/linked_list.h>
+
+/*
+ * A message bus is a mapping of integer message types -> listeners/callbacks.
+ * A listener can listen to a single message, or to all messages on a bus
+ * Message types must be consecutive integers, starting at 1. 0 is reserved
+ * for broadcast to all handlers.
+ * Handlers will be sent a message of type 0, with a NULL msg_data/user_data
+ * when it is time to clean any state up.
+ * Listeners are owned by the subscriber, and are no longer referenced by the bus
+ * once unsubscribed.
+ * Message delivery happens in a separate thread from sending, so listeners are
+ * responsible for their own thread safety
+ */
+struct aws_message_bus;
+
+/* Signature for listener callbacks */
+typedef void(aws_message_bus_listener_fn)(uintptr_t msg_type, void *msg_data, void *user_data);
+
+enum aws_message_bus_policy {
+    /* Message delivery is queued */
+    AWS_BUS_ASYNC,
```

Comment:
super debatable, just throwing the idea out there: Should we just call **everything** `aws_bus_xyz` instead of `aws_message_bus_xyz`?

Reply:
Totally on board with that. Also, do we even like the name bus?

In previous iterations in my career this has been called:
* event_stream (bad idea in this case, overloads that name in our domain)
* message_stream (still not great, because it's not really a serial stream)
* message_queue (correct in the async case, not correct in the sync case)
* event_bus/EBus

DISCUSS!